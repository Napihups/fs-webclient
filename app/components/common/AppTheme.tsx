import React, { ReactNode, useContext, useEffect, useState } from "react";

type Theme = "fsLight" | "fsDark";

type AppThemeContextProps = {
  theme: Theme | null;
  setTheme: (theme: Theme | null) => void;
};

const AppthemeContext = React.createContext<AppThemeContextProps>({
  theme: "fsDark",
  setTheme: () => {},
});

export const useAppTheme = () => {
  return useContext(AppthemeContext);
};

type AppThemeProviderProps = {
  children: ReactNode;
};

/**
 * Set the default theme to Dark
 */
const DEFAULT_THEME = "fsDark";

export const AppThemeProvider: React.FC<AppThemeProviderProps> = ({ children }) => {
  const [theme, setTheme] = useState<Theme | null>(null);

  useEffect(() => {
    const html = document.getElementsByTagName("html")[0];
    if (theme === null) {
      /** Scripts for setting initial theme mode and
       * listening to on Change event regarding the theme being set */

      /** if there is no current theme being set, use browser prefered mode */
      const currentThemeCached = window.localStorage.getItem("data-theme");

      if (currentThemeCached === undefined || currentThemeCached === null) {
        if (window.matchMedia && window.matchMedia("(prefers-color-scheme: dark)").matches) {
          window.localStorage.setItem("data-theme", "fsDark");
          html.setAttribute("data-theme", "fsDark");
          setTheme("fsDark");
          return;
        }
        window.localStorage.setItem("data-theme", "fsLight");
        html.setAttribute("data-theme", "fsLight");
        setTheme("fsLight");
      } else {
        html.setAttribute("data-theme", currentThemeCached);
        setTheme(currentThemeCached as Theme);
      }
    } else {
      window.localStorage.setItem("data-theme", theme);
      html.setAttribute("data-theme", theme);
    }
  }, [theme]);

  const value = {
    theme,
    setTheme,
  };

  return <AppthemeContext.Provider value={value}>{children}</AppthemeContext.Provider>;
};
