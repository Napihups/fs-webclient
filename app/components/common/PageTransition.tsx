import { PageLoader } from "@element/PageLoader/PageLoader";
import { useRouter } from "next/router";
import React, { useEffect, useState } from "react";

export const PageTransition: React.FC = () => {
  const router = useRouter();

  const [state, setState] = useState({
    isRouteChanging: false,
    loadingKey: 0,
  });

  useEffect(() => {
    const handleRouteChangeStart = () => {
      setState((prevState) => ({
        ...prevState,
        isRouteChanging: true,
        loadingKey: prevState.loadingKey ^ 1,
      }));
    };

    const handleRouteChangeEnd = () => {
      setState((prevState) => ({
        ...prevState,
        isRouteChanging: false,
      }));
    };

    router.events.on("routeChangeStart", handleRouteChangeStart);
    router.events.on("routeChangeComplete", handleRouteChangeEnd);
    router.events.on("routeChangeError", handleRouteChangeEnd);

    return () => {
      router.events.off("routeChangeStart", handleRouteChangeStart);
      router.events.off("routeChangeComplete", handleRouteChangeEnd);
      router.events.off("routeChangeError", handleRouteChangeEnd);
    };
  }, [router.events]);

  return <PageLoader isRouteChanging={state.isRouteChanging} key={state.loadingKey} />;
};
