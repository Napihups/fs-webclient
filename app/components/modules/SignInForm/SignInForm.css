.signInForm {
  @apply h-[auto]
  w-full
  flex 
  flex-col
  m-auto;
}

.signInForm__title {
  @apply text-base-content
  text-lg
  font-normal;
}

.signInForm__formControl {
  @apply form-control w-full p-0 my-2;
}

.signInForm__showPassword {
  @apply form-control flex flex-row items-center w-full justify-end mb-4;
}
.signInForm__formError {
  @apply alert bg-transparent text-error my-2 pl-0 py-1 flex flex-row justify-center;
}
