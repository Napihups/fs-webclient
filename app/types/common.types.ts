export enum FormStatus {
  IDLE = "IDLE",
  SUBMITTING = "SUBMITTING",
  ONERROR = "ONERROR",
  ONSUCCESS = "ONSUCCESS",
}
