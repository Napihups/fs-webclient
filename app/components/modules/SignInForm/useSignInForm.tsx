import { FormStatus } from "@fsTypes/common.types";
import { SignInFormSchema } from "@schemas/signin-form.schema";
import React, { useState } from "react";

export const useSignInForm = () => {
  const [emailInputError, setEmailInputError] = useState<string | null>(null);
  const [passwordInputError, setPasswordInputError] = useState<string | null>(null);
  const [formError, setFormError] = useState<string | null>(null);
  const [formStatus, setFormStatus] = useState<FormStatus>(FormStatus.IDLE);

  const submit = async (ev: React.FormEvent) => {
    ev.preventDefault();

    const email = document.getElementById("signin-email") as HTMLInputElement;
    const password = document.getElementById("signin-password") as HTMLInputElement;

    email.blur();
    password.blur();
    setFormError(null);
    const { error, value } = SignInFormSchema.validate({
      email: email.value,
      password: password.value,
    });

    const canSubmit = Boolean(formStatus === FormStatus.IDLE || formStatus === FormStatus.ONERROR);

    if (error === undefined && canSubmit) {
      /**
       * Auth Signin Form are safe to submit
       */
      setFormStatus(FormStatus.SUBMITTING);

      setTimeout(() => {
        setFormStatus(FormStatus.ONERROR);
        setFormError("Invalid Email / Password");
      }, 3000);
    } else {
      /**
       * Auth Signin Form are not safe to submit
       */
      const err = error?.details[0];
      const errPath = err?.path[0] as string;

      switch (errPath) {
        case "email":
          setEmailInputError(err?.message as string);
          break;
        case "password":
          setPasswordInputError(err?.message as string);
          break;
      }
      setFormStatus(FormStatus.ONERROR);
    }
  };

  return {
    emailInputError,
    setEmailInputError,
    passwordInputError,
    setPasswordInputError,
    formError,
    formStatus,
    submit,
  };
};
