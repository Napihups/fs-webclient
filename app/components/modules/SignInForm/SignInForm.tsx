import { FIXTURES } from "@constant/fixtures";
import React, { useState, useRef, useEffect } from "react";
import { useSignInForm } from "./useSignInForm";
import { BiLoaderAlt as Spinner } from "react-icons/bi";
import { IoAlertCircleOutline as Alert } from "react-icons/io5";
import { FormStatus } from "@fsTypes/common.types";

export const SignInForm: React.FC = () => {
  const emailInputRef = useRef<HTMLInputElement>(null);
  const passwordInputRef = useRef<HTMLInputElement>(null);
  const [showPassword, setShowPassword] = useState<boolean>(false);

  useEffect(() => {
    emailInputRef.current?.focus();
  }, [emailInputRef]);

  const form = useSignInForm();

  const onShowPasswordChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    setShowPassword(ev.target.checked);
  };

  const onEmailInputBlur = (ev: React.FocusEvent) => {
    const value = emailInputRef.current?.value ?? undefined;
    if (value !== undefined && value.length > 0) {
      form.setEmailInputError(null);
    }
  };

  const onPasswordInputBlur = (ev: React.FocusEvent) => {
    const value = passwordInputRef.current?.value ?? undefined;
    if (value !== undefined && value.length > 0) {
      form.setPasswordInputError(null);
    }
  };

  return (
    <form className="signInForm" role="form" onSubmit={form.submit}>
      <p className="signInForm__title">{FIXTURES.signinPage.signInFormTitle}</p>

      <div className="signInForm__formControl">
        <input
          id="signin-email"
          role="textbox"
          autoFocus={true}
          ref={emailInputRef}
          onBlur={onEmailInputBlur}
          autoComplete="off"
          className={`input input-bordered ${form.emailInputError !== null ? "input-error" : ""}`}
          placeholder={FIXTURES.signinPage.signInForm.emailPlaceholder}
        />
        {form.emailInputError !== null && (
          <label className="label">
            <span className="label-text-alt text-error">{form.emailInputError}</span>
          </label>
        )}
      </div>
      <div className="signInForm__formControl">
        <input
          role="textbox"
          ref={passwordInputRef}
          onBlur={onPasswordInputBlur}
          id="signin-password"
          className={`input input-bordered ${
            form.passwordInputError !== null ? "input-error" : ""
          }`}
          type={showPassword ? "text" : "password"}
          placeholder={FIXTURES.signinPage.signInForm.passwordPlaceholder}
        />
        {form.passwordInputError !== null && (
          <label className="label">
            <span className="label-text-alt text-error">{form.passwordInputError}</span>
          </label>
        )}
      </div>

      <div className="signInForm__showPassword">
        <input
          role="checkbox"
          id="signin-email"
          className="checkbox"
          type="checkbox"
          onChange={onShowPasswordChange}
        />
        <label className="label text-sm">{FIXTURES.signinPage.signInForm.showPasswordLabel}</label>
      </div>

      {form.formError !== null && (
        <div className="signInForm__formError">
          <div>
            <Alert size={20} />
            <p className="text-center">{form.formError}</p>
          </div>
        </div>
      )}
      <div className="grid grid-rows-2 gap-2">
        <button type="submit" className="btn btn-primary">
          {form.formStatus === FormStatus.SUBMITTING ? (
            <Spinner
              size={30}
              data-testid="signin-loader"
              role="progressbar"
              className="animate-spin"
            />
          ) : (
            <>{FIXTURES.signinPage.signInForm.signInButtonText}</>
          )}
        </button>
        <button type="button" className="btn btn-ghost">
          {FIXTURES.signinPage.signInForm.signUpButtonText}
        </button>
      </div>
    </form>
  );
};
