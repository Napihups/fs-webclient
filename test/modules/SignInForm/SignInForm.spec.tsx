import { FormStatus } from "@fsTypes/common.types";
import { SignInForm } from "@module/SignInForm/SignInForm";
import { screen, render, waitFor, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { not } from "joi";

declare global {
  interface Window {
    _virtualConsole: any;
  }
}

describe("SignInForm UI Test", () => {
  let emit: any;

  beforeAll(() => {
    ({ emit } = window._virtualConsole);
  });

  beforeEach(() => {
    window._virtualConsole.emit = jest.fn();
    jest.clearAllMocks();
  });

  afterEach(() => {
    cleanup();
  });

  afterAll(() => {
    window._virtualConsole.emit = emit;
  });

  test("Static: All UI static textual should be as per requirement & snapshot", async () => {
    const comp = render(<SignInForm />);

    await screen.findByText("Sign in");
    await screen.findByPlaceholderText("Email Address");
    await screen.findByText("Show password");
    await screen.findByText("Login");
    await screen.findByText("Sign Up");

    expect(comp).toMatchSnapshot();
  });

  test("Static: On initial render, the email input should be focus automatically", async () => {
    render(<SignInForm />);

    const emailInput = await screen.findByPlaceholderText("Email Address");
    expect(emailInput).toHaveFocus();
  });

  test("UseCase: Password should use normal text when the Show Password checkbox is checked", async () => {
    const user = userEvent.setup();

    render(<SignInForm />);

    const passwordInput = await screen.findByPlaceholderText("Password");
    const checkbox = await screen.findByRole("checkbox");
    await user.click(checkbox);

    expect(passwordInput).toHaveAttribute("type");
    expect(passwordInput.getAttribute("type")).toBe("text");

    await user.click(checkbox);
    expect(passwordInput.getAttribute("type")).toBe("password");
  });

  test(`UseCase: User typing invalid email and 
  click on Login button should display Email 
  input error with message "Invalid email address"`, async () => {
    const user = userEvent.setup();

    render(<SignInForm />);
    const emailInput = await screen.findByPlaceholderText("Email Address");
    const loginButton = await screen.findByText("Login");

    await user.type(emailInput, "napif4578");

    await user.click(loginButton);

    await new Promise(process.nextTick);

    const target = screen.getByText("Invalid email address");
    expect(target).toBeInTheDocument();
  });

  test(`Usecase : User Typing valid email and invalid password follow by click on login, will
  display password input Error with message of "Please provide a valid password"`, async () => {
    const user = userEvent.setup();

    render(<SignInForm />);
    const emailInput = await screen.findByPlaceholderText("Email Address");
    const passwordInput = await screen.findByPlaceholderText("Password");
    const loginButton = await screen.findByText("Login");

    await user.type(emailInput, "napi@gmail.com");
    await user.type(passwordInput, "lala");

    await user.click(loginButton);

    await new Promise(process.nextTick);

    const target = screen.getByText("Please provide a valid password");
    expect(target).toBeInTheDocument();
  });

  test("Usecase : User enter both valid email and password, signin form should be able to be in submit status", async () => {
    const user = userEvent.setup();

    render(<SignInForm />);
    const emailInput = await screen.findByPlaceholderText("Email Address");
    const passwordInput = await screen.findByPlaceholderText("Password");
    const loginButton = await screen.findByText("Login");

    await user.type(emailInput, "napi@gmail.com");
    await user.type(passwordInput, "Sayang_Afifah07");

    await user.click(loginButton);

    await new Promise(process.nextTick);

    const loader = await screen.findByTestId("signin-loader");
    expect(loader).toBeInTheDocument();
  });

  test("Usecase : User press tab, should autofocus on the next element which is the password input", async () => {
    const user = userEvent.setup();
    render(<SignInForm />);

    const emailInput = await screen.findByPlaceholderText("Email Address");
    const passwordInput = await screen.findByPlaceholderText("Password");

    await user.type(emailInput, "napi@gmail.com");
    await user.type(passwordInput, "Sayang_Afifah07");

    await user.keyboard("{Enter}");

    expect(await screen.getByTestId("signin-loader")).toBeDefined();
  });

  test("Usecase : User after filling the signin form, pressing enter key should allow form to be in submit state", async () => {
    const user = userEvent.setup();
    render(<SignInForm />);

    const passwordInput = await screen.findByPlaceholderText("Password");

    await user.keyboard("{Tab}");
    expect(passwordInput).toHaveFocus();
  });

  test("Behaviour: Email Input should display error when emailInputError state have value", async () => {
    const emailErrorMesage = "This field is required";
    const useSignInForm = jest.spyOn(require("@module/SignInForm/useSignInForm"), "useSignInForm");
    useSignInForm.mockImplementation(() => {
      return {
        emailInputError: emailErrorMesage,
        setEmailInputError: () => {},
        passwordInputError: null,
        setPasswordInputError: () => {},
        formError: null,
        formStatus: null,
        submit: () => {},
      };
    });
    render(<SignInForm />);
    const emailInput = await screen.findByPlaceholderText("Email Address");
    const errElement = await screen.findByText(emailErrorMesage);
    expect(errElement).toBeVisible();
    expect(emailInput).toHaveClass("input-error");

    useSignInForm.mockClear();
  });

  test("Behaviour: Password Input Should display error when passwordInputError state have value", async () => {
    const passwordErrorMessage = "Password is invalid";
    const useSignInForm = jest.spyOn(require("@module/SignInForm/useSignInForm"), "useSignInForm");
    useSignInForm.mockImplementation(() => {
      return {
        emailInputError: null,
        setEmailInputError: () => {},
        passwordInputError: passwordErrorMessage,
        setPasswordInputError: () => {},
        formError: null,
        formStatus: null,
        submit: () => {},
      };
    });

    render(<SignInForm />);
    const passwordInput = await screen.findByPlaceholderText("Password");
    const errElement = await screen.findByText(passwordErrorMessage);
    expect(errElement).toBeVisible();
    expect(passwordInput).toHaveClass("input-error");

    useSignInForm.mockClear();
  });

  test("Behaviour: Form should display FormError text when formError state have value", async () => {
    const formErrorMessage = "Invalid Email / Password";
    const useSignInForm = jest.spyOn(require("@module/SignInForm/useSignInForm"), "useSignInForm");
    useSignInForm.mockImplementation(() => {
      return {
        emailInputError: null,
        setEmailInputError: () => {},
        passwordInputError: null,
        setPasswordInputError: () => {},
        formError: formErrorMessage,
        formStatus: null,
        submit: () => {},
      };
    });

    render(<SignInForm />);
    const errElement = await screen.findByText(formErrorMessage);
    expect(errElement).toBeVisible();

    useSignInForm.mockClear();
  });

  test('Behaviour: Form submit button "login" should display loader during fomr onsubmit progress state', async () => {
    const useSignInForm = jest.spyOn(require("@module/SignInForm/useSignInForm"), "useSignInForm");
    useSignInForm.mockImplementation(() => {
      return {
        emailInputError: null,
        setEmailInputError: () => {},
        passwordInputError: null,
        setPasswordInputError: () => {},
        formError: null,
        formStatus: FormStatus.SUBMITTING,
        submit: () => {},
      };
    });
    render(<SignInForm />);

    await screen.findByTestId("signin-loader");
  });
});
