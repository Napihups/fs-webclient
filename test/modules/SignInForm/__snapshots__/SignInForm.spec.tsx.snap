// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm UI Static & Behaviour Test Static: All UI static textual should be as per requirement & snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="signInForm"
        role="form"
      >
        <p
          class="signInForm__title"
        >
          Sign in
        </p>
        <div
          class="signInForm__formControl"
        >
          <input
            autocomplete="off"
            class="input input-bordered "
            id="signin-email"
            placeholder="Email Address"
            role="textbox"
          />
        </div>
        <div
          class="signInForm__formControl"
        >
          <input
            class="input input-bordered "
            id="signin-password"
            placeholder="Password"
            role="textbox"
            type="password"
          />
        </div>
        <div
          class="signInForm__showPassword"
        >
          <input
            class="checkbox"
            id="signin-email"
            role="checkbox"
            type="checkbox"
          />
          <label
            class="label text-sm"
          >
            Show password
          </label>
        </div>
        <div
          class="grid grid-rows-2 gap-2"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            Login
          </button>
          <button
            class="btn btn-ghost"
            type="button"
          >
            Sign Up
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="signInForm"
      role="form"
    >
      <p
        class="signInForm__title"
      >
        Sign in
      </p>
      <div
        class="signInForm__formControl"
      >
        <input
          autocomplete="off"
          class="input input-bordered "
          id="signin-email"
          placeholder="Email Address"
          role="textbox"
        />
      </div>
      <div
        class="signInForm__formControl"
      >
        <input
          class="input input-bordered "
          id="signin-password"
          placeholder="Password"
          role="textbox"
          type="password"
        />
      </div>
      <div
        class="signInForm__showPassword"
      >
        <input
          class="checkbox"
          id="signin-email"
          role="checkbox"
          type="checkbox"
        />
        <label
          class="label text-sm"
        >
          Show password
        </label>
      </div>
      <div
        class="grid grid-rows-2 gap-2"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Login
        </button>
        <button
          class="btn btn-ghost"
          type="button"
        >
          Sign Up
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
